AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientSignInTests.swift

```diff
@@ -125,6 +125,66 @@ class AWSMobileClientSignInTests: AWSMobileClientTestBase {
                      "Current sign callback should be nil")
     }
     
+    /// Test whether the confirm sign in completion source is reset when complete
+    ///
+    /// - Given: An unauthenticated session with MFA
+    /// - When:
+    ///    - I invoke `signIn` with completion callback
+    ///    - And then invoke `confirmSignIn` from the above callback
+    ///    - And then I wait for completion
+    /// - Then:
+    ///    - My `confirmSignIn` completion source should be nil
+    ///
+    func testConfirmSignCompletionSourceReset() {
+        let username = "testUser" + UUID().uuidString
+        adminCreateUser(username: username,
+                        temporaryPassword: AWSMobileClientTestBase.sharedPassword,
+                        userAttributes: [
+                            "email": AWSMobileClientTestBase.sharedEmail
+                        ])
+        let confirmSignInWasSuccessfulExpectation = expectation(description: "confirmSignIn was successful")
+        
+        AWSMobileClient.default().signIn(username: username,
+                                         password: AWSMobileClientTestBase.sharedPassword) { (signInResult, error) in
+            if let error = error {
+                XCTFail("User login failed: \(error.localizedDescription)")
+                return
+            }
+            
+            guard let signInResult = signInResult else {
+                XCTFail("User login failed, signInResult unexpectedly nil")
+                return
+            }
+            
+            XCTAssert(signInResult.signInState == .newPasswordRequired)
+            
+            AWSMobileClient.default().confirmSignIn(
+                challengeResponse: AWSMobileClientTestBase.sharedPassword
+            ) { (signInResult, error) in
+                defer {
+                    confirmSignInWasSuccessfulExpectation.fulfill()
+                }
+                
+                if let error = error {
+                    XCTFail("Confirm sign in failed: \(error.localizedDescription)")
+                    return
+                }
+                
+                guard let signInResult = signInResult else {
+                    XCTFail("Confirm sign in failed, signInResult unexpectedly nil")
+                    return
+                }
+                
+                XCTAssert(signInResult.signInState == .signedIn)
+                XCTAssertNil(AWSMobileClient.default().userpoolOpsHelper.currentConfirmSignInHandlerCallback,
+                                "Current confirmsign callback should be nil")
+                XCTAssertNil(AWSMobileClient.default().userpoolOpsHelper.newPasswordRequiredTaskCompletionSource,
+                                "New password completion source should be nil")
```

Comment:
instead of just checking the values of the those tasks, can we also verify that calling `AWSMobileClient.default().confirmSignIn` twice doesn't cause a crash anymore?

Reply:
Absolutely, and good thinking. I've updated the test to perform the createUser/confirmSignIn flow twice, which is the use case for which we experienced the crash originally. Thanks!